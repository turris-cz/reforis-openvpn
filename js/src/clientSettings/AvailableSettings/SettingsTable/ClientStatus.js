/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";

import { faRedo } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { useAPIGet, Button, SpinnerElement, API_STATE } from "foris";
import PropTypes from "prop-types";

import API_URLs from "API";

ClientStatus.propTypes = {
    id: PropTypes.string.isRequired,
    defaultStatus: PropTypes.bool.isRequired,
};

export default function ClientStatus({ id, defaultStatus }) {
    // No need to worry about default state from props as any update will reload whole table
    const [status, setStatus] = useState(defaultStatus);
    const [response, getSettings] = useAPIGet(
        `${API_URLs.clientSettings}/${id}`
    );

    useEffect(() => {
        if (response.state === API_STATE.SUCCESS) {
            setStatus(response.data.running);
        }
    }, [response]);

    // GET is sent on demand hence INIT is not checked
    if (response.state === API_STATE.SENDING) {
        return <SpinnerElement small />;
    }

    if (response.state === API_STATE.ERROR) {
        return (
            <span className="text-danger">{_("Cannot refresh status")}</span>
        );
    }

    return (
        <>
            {status ? (
                _("Running")
            ) : (
                <span className="text-danger">{_("Not running")}</span>
            )}
            <Button
                className="btn btn-sm"
                aria-label={_("Check status")}
                onClick={getSettings}
                title={_("Check status")}
            >
                <FontAwesomeIcon icon={faRedo} />
            </Button>
        </>
    );
}
