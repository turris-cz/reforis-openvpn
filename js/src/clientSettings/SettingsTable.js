/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useState } from "react";
import PropTypes from "prop-types";
import { useUID } from "react-uid";

import {
    useAPIGet, useAPIDelete, formFieldsSize, useAlert, Button, SpinnerElement,
    API_STATE, withErrorMessage, withSpinnerOnSending, useAPIPut,
} from "foris";

import API_URLs from "API";

export default function SettingsTable() {
    const [getSettingsResponse, getSettings] = useAPIGet(API_URLs.clientSettings);
    useEffect(() => {
        getSettings();
    }, [getSettings]);

    return (
        <>
            <h3>{_("Available settings")}</h3>
            <p>{_("For each uploaded file a new OpenVPN client instance is created. Please check settings file for errors if instance is enabled and not running few minutes after setup.")}</p>
            <ClientTableWithErrorAndSpinner
                apiState={getSettingsResponse.state}
                settingsData={getSettingsResponse.data}
            />
        </>
    );
}

const ClientTableWithErrorAndSpinner = withErrorMessage(withSpinnerOnSending(ClientTable));

ClientTable.propTypes = {
    settingsData: PropTypes.array.isRequired,
};

function ClientTable({ settingsData }) {
    if (!settingsData || settingsData.length === 0) {
        return <p className="text-muted text-center">{_("There are no settings added yet.")}</p>;
    }

    return (
        <table className={`table table-hover mt-3 ${formFieldsSize}`}>
            <tbody>
                {settingsData.map((client) => <ClientRow key={client.id} client={client} />)}
            </tbody>
        </table>
    );
}

ClientRow.propTypes = {
    client: PropTypes.shape({
        id: PropTypes.string.isRequired,
        enabled: PropTypes.bool.isRequired,
        running: PropTypes.bool.isRequired,
    }).isRequired,
};

function ClientRow({ client }) {
    const uid = useUID();
    const [setAlert] = useAlert();

    const [putClientResponse, putClient] = useAPIPut(`${API_URLs.clientSettings}/${client.id}`);
    useEffect(() => {
        if (putClientResponse.state === API_STATE.ERROR) {
            setAlert(putClientResponse.data);
        }
    }, [putClientResponse, setAlert]);

    const [deleteClientResponse, deleteClient] = useAPIDelete(`${API_URLs.clientSettings}/${client.id}`);
    useEffect(() => {
        if (deleteClientResponse.state === API_STATE.ERROR) {
            setAlert(deleteClientResponse.data);
        }
    }, [deleteClientResponse, setAlert]);

    return (
        <tr>
            <td className="align-middle">
                <div className="custom-control custom-switch">
                    <input
                        type="checkbox"
                        className="custom-control-input"
                        id={uid}
                        defaultChecked={client.enabled}
                        onChange={(event) => putClient({ enabled: event.target.checked })}
                    />
                    <label className="custom-control-label" htmlFor={uid}>{client.id}</label>
                </div>
            </td>
            <td className="align-middle text-center">
                {client.enabled && <ClientStatus id={client.id} defaultStatus={client.running} />}
            </td>
            <td className="text-right">
                <button type="button" className="btn btn-danger" onClick={deleteClient}>
                    <i className="fa fa-trash-alt mr-2" />
                    {_("Delete")}
                </button>
            </td>
        </tr>
    );
}

ClientStatus.propTypes = {
    id: PropTypes.string.isRequired,
    defaultStatus: PropTypes.bool.isRequired,
};

function ClientStatus({ id, defaultStatus }) {
    // No need to worry about default state from props as any update will reload whole table
    const [status, setStatus] = useState(defaultStatus);

    const [response, getSettings] = useAPIGet(`${API_URLs.clientSettings}/${id}`);
    useEffect(() => {
        if (response.state === API_STATE.SUCCESS) {
            setStatus(response.data.running);
        }
    }, [response]);

    // GET is sent on demand hence INIT is not checked
    if (response.state === API_STATE.SENDING) {
        return <SpinnerElement />;
    }
    if (response.state === API_STATE.ERROR) {
        return <span className="text-danger">{_("Cannot refresh status")}</span>;
    }

    return (
        <>
            {status ? _("Running") : <span className="text-danger">{_("Not running")}</span>}
            <Button className="btn" aria-label={_("Check status")} onClick={getSettings}>
                <i className="fa fa-redo" />
            </Button>
        </>
    );
}
