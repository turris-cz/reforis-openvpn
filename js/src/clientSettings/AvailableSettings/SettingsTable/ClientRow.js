/*
 * Copyright (C) 2019-2024 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import { faEdit, faTrash } from "@fortawesome/free-solid-svg-icons";
import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import {
    useAPIDelete,
    useAlert,
    Button,
    API_STATE,
    useAPIPut,
    Switch,
} from "foris";
import PropTypes from "prop-types";

import API_URLs from "API";

import ClientStatus from "./ClientStatus";

ClientRow.propTypes = {
    client: PropTypes.shape({
        id: PropTypes.string.isRequired,
        enabled: PropTypes.bool.isRequired,
        running: PropTypes.bool.isRequired,
    }).isRequired,
    editClientSettings: PropTypes.func.isRequired,
};

export default function ClientRow({ client, editClientSettings }) {
    const [setAlert] = useAlert();

    const [putClientResponse, putClient] = useAPIPut(
        `${API_URLs.clientSettings}/${client.id}`
    );

    useEffect(() => {
        if (putClientResponse.state === API_STATE.ERROR) {
            setAlert(putClientResponse.data);
        }
    }, [putClientResponse, setAlert]);

    const [deleteClientResponse, deleteClient] = useAPIDelete(
        `${API_URLs.clientSettings}/${client.id}`
    );

    useEffect(() => {
        if (deleteClientResponse.state === API_STATE.ERROR) {
            setAlert(deleteClientResponse.data);
        }
    }, [deleteClientResponse, setAlert]);

    const buttonDisabled =
        deleteClientResponse.state === API_STATE.SENDING ||
        putClientResponse.state === API_STATE.SENDING;

    return (
        <tr>
            <td className="align-middle">
                <Switch
                    label={client.id}
                    checked={client.enabled}
                    onChange={(event) => {
                        putClient({
                            data: { enabled: event.target.checked },
                        });
                    }}
                    disabled={buttonDisabled}
                    className="mb-0"
                />
            </td>
            <td className="align-middle text-center">
                {client.enabled && (
                    <ClientStatus
                        id={client.id}
                        defaultStatus={client.running}
                    />
                )}
            </td>
            <td className="text-end">
                <ClientSettingsActions
                    client={client}
                    editClientSettings={editClientSettings}
                    deleteClient={deleteClient}
                    disabled={buttonDisabled}
                />
            </td>
        </tr>
    );
}

ClientSettingsActions.propTypes = {
    client: PropTypes.object,
    editClientSettings: PropTypes.func,
    deleteClient: PropTypes.func,
    disabled: PropTypes.bool,
};

function ClientSettingsActions({
    client,
    editClientSettings,
    deleteClient,
    disabled,
}) {
    return (
        <div
            className="btn-group btn-group-sm mb-0"
            role="group"
            aria-label={_("Actions")}
        >
            <Button onClick={() => editClientSettings(client)}>
                <span className="d-xl-none">
                    <FontAwesomeIcon icon={faEdit} className="fa-sm" />
                </span>
                <span className="d-none d-xl-block">
                    <FontAwesomeIcon icon={faEdit} className="fa-sm me-1" />
                    {_("Edit")}
                </span>
            </Button>
            <Button
                onClick={deleteClient}
                className="btn-danger"
                disabled={disabled}
            >
                <span className="d-xl-none">
                    <FontAwesomeIcon icon={faTrash} className="fa-sm" />
                </span>
                <span className="d-none d-xl-block">
                    <FontAwesomeIcon icon={faTrash} className="fa-sm me-1" />
                    {_("Delete")}
                </span>
            </Button>
        </div>
    );
}
