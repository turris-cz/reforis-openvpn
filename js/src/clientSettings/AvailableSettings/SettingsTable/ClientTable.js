/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";

import PropTypes from "prop-types";

import ClientRow from "./ClientRow";

ClientTable.propTypes = {
    clientSettings: PropTypes.array.isRequired,
    editClientSettings: PropTypes.func.isRequired,
};

export default function ClientTable({ clientSettings, editClientSettings }) {
    if (!clientSettings || clientSettings.length === 0) {
        return (
            <p className="text-muted text-center">
                {_("There are no settings added yet.")}
            </p>
        );
    }

    return (
        <div className="table-responsive">
            <table className="table table-hover mb-0">
                <thead className="thead-light">
                    <tr>
                        <th>{_("Instances")}</th>
                        <th
                            scope="col"
                            aria-label={_("Actions")}
                            className="text-center"
                        />
                        <th
                            scope="col"
                            aria-label={_("Actions")}
                            className="text-end"
                        />
                    </tr>
                </thead>
                <tbody>
                    {clientSettings.map((client) => (
                        <ClientRow
                            key={client.id}
                            client={client}
                            editClientSettings={editClientSettings}
                        />
                    ))}
                </tbody>
            </table>
        </div>
    );
}
