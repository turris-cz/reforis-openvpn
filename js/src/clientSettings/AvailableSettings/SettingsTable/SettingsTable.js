/*
 * Copyright (C) 2019-2022 CZ.NIC z.s.p.o. (https://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useState } from "react";

import PropTypes from "prop-types";

import ClientTable from "./ClientTable";
import SettingsModal from "../SettingsModal/SettingsModal";

SettingsTable.propTypes = {
    clientSettings: PropTypes.arrayOf(
        PropTypes.shape({
            id: PropTypes.string.isRequired,
            enabled: PropTypes.bool.isRequired,
            running: PropTypes.bool.isRequired,
            credentials: PropTypes.shape({
                username: PropTypes.string.isRequired,
                password: PropTypes.string.isRequired,
            }),
        })
    ).isRequired,
};

export default function SettingsTable({ clientSettings }) {
    const [clientToEdit, setClientToEdit] = useState(null);
    const [isClientModalShown, setIsClientModalShown] = useState(false);

    const handleEditClientSettings = (client) => {
        setClientToEdit(client);
        setIsClientModalShown(true);
    };

    return (
        <>
            <ClientTable
                clientSettings={clientSettings}
                editClientSettings={handleEditClientSettings}
            />
            <SettingsModal
                shown={isClientModalShown}
                setShown={setIsClientModalShown}
                client={clientToEdit}
            />
        </>
    );
}
