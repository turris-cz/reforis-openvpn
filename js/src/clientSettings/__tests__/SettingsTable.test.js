/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    render, fireEvent, wait, getByText, waitForElement, getByLabelText,
} from "customTestRender";
import mockAxios from "jest-mock-axios";

import { AlertContext } from "foris";

import SettingsTable from "../SettingsTable";

describe("<SettingsTable />", () => {
    const setAlert = jest.fn();
    const singleClient = { id: "A1", enabled: true, running: true };

    async function renderTable(clients = []) {
        const { container } = render(
            <AlertContext.Provider value={setAlert}>
                <SettingsTable />
            </AlertContext.Provider>,
        );
        mockAxios.mockResponse({ data: clients });
        await waitForElement(() => getByText(container, "Available settings"));
        return container;
    }

    function getDelete(container) {
        return getByText(container, "Delete");
    }

    it("should display spinner", () => {
        const { container } = render(<SettingsTable />);
        expect(container).toMatchSnapshot();
    });

    it("should handle GET error", async () => {
        const { container } = render(<SettingsTable />);
        mockAxios.mockError({response: {}});
        await waitForElement(() => getByText(container, "An error occurred during loading OpenVPN client settings"));
    });

    it("should render info when there are no clients", async () => {
        const container = await renderTable();
        expect(container).toMatchSnapshot();
    });

    it("should render table of settings", async () => {
        const settings = [
            { id: "A1", enabled: true, running: true },
            { id: "A2", enabled: false, running: false },
            { id: "B1", enabled: true, running: true },
            { id: "B2", enabled: false, running: false },
        ];
        const container = await renderTable(settings);
        expect(container).toMatchSnapshot();
    });

    it("should send PATCH request", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getByText(container, singleClient.id));
        expect(mockAxios.patch).toBeCalledWith(
            `/reforis/openvpn/api/client-settings/${singleClient.id}`,
            { enabled: false },
            expect.anything(),
        );
    });

    it("should handle PATCH error", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getByText(container, singleClient.id));
        const errorMessage = "API didn't handle this well";
        mockAxios.mockError(
            { response: { data: errorMessage, headers: { "content-type": "application/json" } } },
        );
        await wait(() => {
            expect(setAlert).toHaveBeenCalledWith(errorMessage);
        });
    });

    it("should send DELETE request", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getDelete(container));
        expect(mockAxios.delete).toBeCalledWith(
            `/reforis/openvpn/api/client-settings/${singleClient.id}`,
            expect.anything(),
        );
    });

    it("should handle DELETE error", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getDelete(container));
        const errorMessage = "API didn't handle this well";
        mockAxios.mockError(
            { response: { data: errorMessage, headers: { "content-type": "application/json" } } },
        );
        await wait(() => {
            expect(setAlert).toHaveBeenCalledWith(errorMessage);
        });
    });

    it("should change status of client instance after refresh", async () => {
        const container = await renderTable([singleClient]);
        expect(getByText(container, "Running")).toBeTruthy();
        fireEvent.click(getByLabelText(container, "Check status"));
        expect(mockAxios.get).toHaveBeenCalledWith(`/reforis/openvpn/api/client-settings/${singleClient.id}`, expect.anything());
        mockAxios.mockResponse({ data: { ...singleClient, running: false }})
        await waitForElement(() => getByText(container, "Not running"));
    });

    it("should change status of client instance after refresh", async () => {
        const container = await renderTable([singleClient]);
        fireEvent.click(getByLabelText(container, "Check status"));
        expect(mockAxios.get).toHaveBeenCalledWith(`/reforis/openvpn/api/client-settings/${singleClient.id}`, expect.anything());
        const errorMessage = "API didn't handle this well";
        mockAxios.mockError(
            { response: { data: errorMessage, headers: { "content-type": "application/json" } } },
        );
        await waitForElement(() => getByText(container, "Cannot refresh status"));
    });
});
