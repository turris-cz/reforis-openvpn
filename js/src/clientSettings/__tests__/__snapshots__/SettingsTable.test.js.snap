// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SettingsTable /> should display spinner 1`] = `
<div>
  <h2>
    Available Settings
  </h2>
  <p>
    For each uploaded file, a new OpenVPN client instance is created. Please check the settings file for errors if the instance is enabled and not running a few minutes after setup.
  </p>
  <div
    class="spinner-wrapper my-3 text-center"
  >
    <div
      class="spinner-border"
      role="status"
    >
      <span
        class="sr-only"
      />
    </div>
  </div>
</div>
`;

exports[`<SettingsTable /> should render info when there are no clients 1`] = `
<div>
  <h2>
    Available Settings
  </h2>
  <p>
    For each uploaded file, a new OpenVPN client instance is created. Please check the settings file for errors if the instance is enabled and not running a few minutes after setup.
  </p>
  <p
    class="text-muted text-center"
  >
    There are no settings added yet.
  </p>
</div>
`;

exports[`<SettingsTable /> should render table of settings 1`] = `
<div>
  <h2>
    Available Settings
  </h2>
  <p>
    For each uploaded file, a new OpenVPN client instance is created. Please check the settings file for errors if the instance is enabled and not running a few minutes after setup.
  </p>
  <div
    class="table-responsive"
  >
    <table
      class="table table-hover mb-0"
    >
      <thead
        class="thead-light"
      >
        <tr>
          <th>
            Instances
          </th>
          <th
            aria-label="Actions"
            class="text-center"
            scope="col"
          />
          <th
            aria-label="Actions"
            class="text-right"
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="align-middle"
          >
            <div
              class="custom-control custom-switch"
            >
              <input
                checked=""
                class="custom-control-input"
                id="1"
                type="checkbox"
              />
              <label
                class="custom-control-label"
                for="1"
              >
                A1
              </label>
            </div>
          </td>
          <td
            class="align-middle text-center"
          >
            Running
            <button
              aria-label="Check status"
              class="btn btn btn-sm"
              title="Check status"
              type="button"
            >
              <i
                class="fa fa-redo"
              />
            </button>
          </td>
          <td
            class="text-right"
          >
            <button
              class="btn btn-sm btn-danger"
              type="button"
            >
              <i
                class="fa fa-trash-alt mr-2"
              />
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            <div
              class="custom-control custom-switch"
            >
              <input
                class="custom-control-input"
                id="2"
                type="checkbox"
              />
              <label
                class="custom-control-label"
                for="2"
              >
                A2
              </label>
            </div>
          </td>
          <td
            class="align-middle text-center"
          />
          <td
            class="text-right"
          >
            <button
              class="btn btn-sm btn-danger"
              type="button"
            >
              <i
                class="fa fa-trash-alt mr-2"
              />
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            <div
              class="custom-control custom-switch"
            >
              <input
                checked=""
                class="custom-control-input"
                id="3"
                type="checkbox"
              />
              <label
                class="custom-control-label"
                for="3"
              >
                B1
              </label>
            </div>
          </td>
          <td
            class="align-middle text-center"
          >
            Running
            <button
              aria-label="Check status"
              class="btn btn btn-sm"
              title="Check status"
              type="button"
            >
              <i
                class="fa fa-redo"
              />
            </button>
          </td>
          <td
            class="text-right"
          >
            <button
              class="btn btn-sm btn-danger"
              type="button"
            >
              <i
                class="fa fa-trash-alt mr-2"
              />
              Delete
            </button>
          </td>
        </tr>
        <tr>
          <td
            class="align-middle"
          >
            <div
              class="custom-control custom-switch"
            >
              <input
                class="custom-control-input"
                id="4"
                type="checkbox"
              />
              <label
                class="custom-control-label"
                for="4"
              >
                B2
              </label>
            </div>
          </td>
          <td
            class="align-middle text-center"
          />
          <td
            class="text-right"
          >
            <button
              class="btn btn-sm btn-danger"
              type="button"
            >
              <i
                class="fa fa-trash-alt mr-2"
              />
              Delete
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</div>
`;
